src/librustc_typeck/check/op.rs

```diff
@@ -186,6 +186,34 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                             "binary operation `{}` cannot be applied to type `{}`",
                             op.node.as_str(),
                             lhs_ty);
+
+                        match lhs_ty.sty {
+                            TypeVariants::TyRef(_, ref ty_mut) => {
+                                match ty_mut.ty.sty {
+                                    TypeVariants::TyRef(_, ref ty_mut) => {
+                                        match self.lookup_op_method(expr, ty_mut.ty, vec![rhs_ty_var],
```

Comment:
So does this mean `&&T + U` will trigger this note when `T + U` exists? This is a bit unfortunate, since the message recommends a single deref. Perhaps just check if `&T + U` works (i.e., remove only one layer of reference)?